chapters/infrastructure/checkers/reading/read.md

```diff
@@ -1,12 +1,129 @@
+
+In order to work on a submission, students must download the submission template code.
+As part of the template, a script (named `local.sh`) runs the submission checker inside of the container infrastructure, mirroring the checking environment in GitLab.
+The checker script can also be called locally outside of the container.
+
+A learner can upload their source code as part of an archive inside the Moodle assignment activity, and it will launch the checking request.
+The Git-based checking flow requires learners to push changes directly to GitLab, where the changes committed trigger the GitLab CI/CD workflow.
+
+Detailed instructions for learners are part of the [official `vmchecker` Student Handbook](https://github.com/open-education-hub/vmchecker-next/wiki/Student-Handbook).
+
+## Container Infrastructure
+
+The issue with using the same assignment checking script for many students is that they can have different libraries and applications installed on their systems.
+Different environments may lead to certain optimization inconsistencies.
+To mitigate this, we recommend teachers setup the checking infrastructure in `vmchecker` so that the scripts run inside of a container.
+The `vmchecker` team provides a minimal template that can be expanded based on individual class needs.
+
+The container infrastructure presents an easier approach than the one seen in current solutions, which either require an internet connection for remote checking, or a virtual machine, which uses more resources.
+Using containers allows for easier patches to the environment, for example installing a new library, because the changes will be sent incrementally, the student not needing to download a whole new image.
+
+## Example `vmchecker` Deployments
+
+`vmchecker` has been used extensively for the Operating Systems class at the [National University of Science and Technology POLITEHNICA of Bucharest (UNSTPB)](https://upb.ro/), using the contents of [the `operating-systems` repository](https://github.com/open-education-hub/operating-systems).
+The repositories used for the five assignments, plus a template repository, plus a tutorial assignment can be used as examples of repositories for `vmchecker` deployment:
+
+- [`Assignment: Base`](https://gitlab.cs.pub.ro/operating-systems/assignment-base)
+- [`Assignment: Tutorial`](https://gitlab.cs.pub.ro/operating-systems/assignment-tutorial)
+- [`Assignment: Mini Libc`](https://gitlab.cs.pub.ro/operating-systems/assignment-mini-libc)
+- [`Assignment: Memory Allocator`](https://gitlab.cs.pub.ro/operating-systems/assignment-memory-allocator)
+- [`Assignment: Parallel Graph`](https://gitlab.cs.pub.ro/operating-systems/assignment-parallel-graph)
+- [`Assignment: Mini Shell`](https://gitlab.cs.pub.ro/operating-systems/assignment-mini-shell)
+- [`Assignment: Async Web Server`](https://gitlab.cs.pub.ro/operating-systems/assignment-async-web-server)
```

Comment:
I think these links are pretty pointless overall. They are inaccessible without logging in using the UPB credentials, so only people from UPB will be able to access them.